arraycontext/impl/pytato.py

```diff
@@ -0,0 +1,536 @@
+                                                          output_naming_map,
+                                                          output_template=outputs)
+
+        return self.program_cache[arg_descrs](*args)
+
+
+class PytatoExecutable:
+    def __init__(self, actx, pytato_program, input_id_to_name_in_program,
+                 output_id_to_name_in_program, output_template):
+        self.actx = actx
+        self.pytato_program = pytato_program
+        self.input_id_to_name_in_program = input_id_to_name_in_program
+        self.output_id_to_name_in_program = output_id_to_name_in_program
+        self.output_template = output_template
+
+    def __call__(self, *args):
+        import pytato as pt
+        import pyopencl.array as cla
+        from arraycontext import (is_array_container,
+                                  rec_keyed_map_array_container)
+
+        input_kwargs_to_loopy = {}
+
+        # {{{ extract loopy arguments execute the program
+
+        for pos, arg in enumerate(args):
+            if isinstance(arg, np.number):
+
+                input_kwargs_to_loopy[self.input_id_to_name_in_program[(pos,)]] = (
+                    cla.to_device(self.actx.queue, np.array(arg)))
+            elif is_array_container(arg):
+                def _extract_lpy_kwargs(keys, ary):
+                    if isinstance(ary, pt.array.DataWrapper):
+                        processed_ary = ary.data
+                    elif isinstance(ary, cla.Array):
+                        processed_ary = ary
+                    elif isinstance(ary, pt.Array):
+                        processed_ary = (self.actx.freeze(ary)
+                                         .with_queue(self.actx.queue))
+                    else:
+                        raise TypeError("Expect pt.Array or CL-array, got "
+                                f"{type(ary)}")
+
+                    input_kwargs_to_loopy[
+                        self.input_id_to_name_in_program[(pos,)
+                                                         + keys]] = processed_ary
+                    return ary
+
+                rec_keyed_map_array_container(_extract_lpy_kwargs, arg)
+            else:
+                raise NotImplementedError(type(arg))
+
+        # {{{ the generated program might not have depended on some of the
+        # inputs => do not pass those to the loopy kernel
+
+        input_kwargs_to_loopy = {arg_name: arg
+                                 for arg_name, arg in input_kwargs_to_loopy.items()
+                                 if arg_name in (self.pytato_program
+                                                 .program.default_entrypoint
+                                                 .arg_dict)}
+
+        # }}}
+
+        evt, out_dict = self.pytato_program(queue=self.actx.queue,
+                                            allocator=self.actx.allocator,
+                                            **input_kwargs_to_loopy)
+
+        evt.wait()
```

Comment:
Why the `wait`?

Reply:
For the DAGs I had generally dealt with, they generally had high global temporary variable allocation for each kernel. And so it made sense to flush the queue before enqueuing another kernel. However, looking at `PyOpenCLArrayContext`, I think I should hack-in `wait_event_queue_length`.